File: learnLeetCode_remote/Day25/Part4/lost_player_ids.py
# The Lost Player IDs
# Trick: missing ID = XOR(1..n) XOR XOR(players appearing odd number of times in swap records).
from typing import List, Tuple

def missing_id(n: int, records: List[Tuple[int,int]]) -> int:
    xr = 0
    for i in range(1, n+1):
        xr ^= i
    freq = [0]*(n+1)
    for a,b in records:
        if 1<=a<=n: freq[a] ^= 1
        if 1<=b<=n: freq[b] ^= 1
    for i in range(1, n+1):
        if freq[i]:
            xr ^= i
    return xr

if __name__ == "__main__":
    n=5; records=[[1,2],[3,4],[2,3]]
    print(missing_id(n, records))  # 5

File: learnLeetCode_remote/Day25/Part4/lost_player_ids.cpp
#include <bits/stdc++.h>
using namespace std;

// The Lost Player IDs
// missing = XOR(1..n) XOR XOR(players with odd participation count in swaps)

int missing_id(int n, const vector<pair<int,int>>& records){
    int xr = 0; for(int i=1;i<=n;++i) xr ^= i;
    vector<int> freq(n+1,0);
    for(auto [a,b]: records){ if(1<=a && a<=n) freq[a]^=1; if(1<=b && b<=n) freq[b]^=1; }
    for(int i=1;i<=n;++i) if(freq[i]) xr ^= i;
    return xr;
}

int main(){
    int n=5; vector<pair<int,int>> rec={{1,2},{3,4},{2,3}};
    cout << missing_id(n, rec) << "\n"; // 5
    return 0;
}
